main/src/library/Set.flix

```diff
@@ -328,23 +316,55 @@ namespace Set {
     /// make any guarantees about which mapping will be in the resulting map.
     ///
     pub def toMap(xs: Set[(a, b)]): Map[a, b] =
-        let Set(s) = xs;
-            List.toMap(s)
+        foldRight((x, acc) -> let (k, v) = x; Map.insert(k, v, acc), Map#{}, xs)
 
     ///
     /// Applies `f` to every element of `xs`.
     ///
     /// The function `f` must be impure.
     ///
-    pub def foreach(f: a ~> Unit, xs: Set[a]): Unit & Impure = match xs {
-        case Set(l) => List.foreach(f, l)
+    pub def foreach(f: a ~> Unit, xs: Set[a]): Unit & Impure =
+        let Set(es) = xs;
+        RBTree.foreach((x, _) -> f(x), es)
+
+    ///
+    /// Helper function for `cmp`.
+    ///
+    def __cmpHelper(xs: List[a], ys: List[a]): Int32 = match (xs, ys) {
+        case (z :: zs, w :: ws) =>
+            let cmp = z <=> w;
+            if (cmp == 0) __cmpHelper(zs, ws) else cmp
+        case (_ :: _, Nil) => 1
+        case (Nil, _ :: _) => -1
+        case (Nil, Nil) => 0
     }
 
+    ///
+    /// Returns the result of a three-way comparison between `xs` and `ys`.
+    ///
+    /// Returns n < 0 if `xs` is lexicographically smaller than `ys`.
+    /// Returns n = 0 if `xs` is equal to `ys`.
+    /// Returns n > 0 if `xs` is lexicographically greater than `ys`.
+    ///
+    pub def __cmp(xs: Set[a], ys: Set[a]): Int32 =
+        __cmpHelper(toList(xs), toList(ys))
```

Comment:
Can this not be done better with some form of parallel tree traversal?

Reply:
Perhaps. I would love to avoid the use of List for this, but  I must admit that __cmp is one of the methods i've struggled a bit with.
I think the challenge is that we need a parallel inorder tree traversal for two trees, but the trees can have quite different structures, even though the elements of the first tree might be a subset of the elements in the second.

Reply:
I see, perhaps interesting future work then.